diffedit.md

```diff
@@ -0,0 +1,752 @@
+```
+
+If there are holes in the mask, there are other techniques that can help, such as `binary_fill_holes` from the `scipy.ndimage` package. We'll skip this since our example doesn't have any holes, but it can be another option to improve the mask further. For now, we'll stick with the mask we've created, which now looks like this:
+
+
+```python
+plot_mask(original_image, mask_np)
+```
+
+
+
+
+    <module 'matplotlib.pyplot' from '/opt/conda/lib/python3.10/site-packages/matplotlib/pyplot.py'>
+
+
+
+
+    
+![png](diffedit_files/diffedit_72_1.png)
+    
+
+
+# Applying DiffEdit to our original image
+
+Now we have a mask, we can utilise this to ensure that our image generation of our zebra only affects the sections under the mask. We do this using a standard image to image pipeline for a diffusion model, but with one main change. For each step of the denoising process, instead of removing the noise that our diffusion model predicts in the areas outside the mask, we will replace this with a version of our original image with the correct amount of noise. By doing this at each step of the denoising process, this allows the final generated image within the mask to seamlessly integrate with the original image outside of the mask, as we effectively trick it at each step into believing that it generated the background of the original image. We'll explain exactly how we do this in more detail later.
+
+For now, we want to set up our image to image pipeline in the standard way, with one exception - we want to create two schedulers, one which handles the denoising of the image we are generating as normal, but another which allows us to correctly noise our original image to the correct level, which we then use within the denoising step.
+
+
+```python
+# set how much of the initial image should be replaced by noise at the start
+noise_strength = 0.75
+# set a manual seed to ensure that we get consistent results
+torch.manual_seed(3)
+
+# create schedulers
+denoising_scheduler = init_scheduler(num_inference_steps)
+original_image_scheduler = init_scheduler(num_inference_steps)
+
+# store step information
+total_steps = len(scheduler.timesteps)
+start_step = int(len(scheduler.timesteps)-(num_inference_steps*noise_strength))
+
+# set up callback to see denoising process
+itermediate_latents = []
+callback = (lambda latents: itermediate_latents.append(latents))
+
+# create text embeddings for new_prompt "zebra"
+text_embeddings = create_text_embeddings([new_prompt])
+
+mask = torch.from_numpy(mask_np).type(torch.half).to("cuda")
+```
+
+We use an image to image pipeline to pass in our original image, since we still want our zebra to have a similar pose to the horse which is present in our original image, so we add a lot of noise to the original image (~75%) and then we will use this as the starting point for our image generation of our zebra. We can see what that initial noise looks like below:
+TODO: should we show the latents or the decoded image?
```

Comment:
I'd argue neither. They both look like a poorly tuned analogue TV!

Reply:
You're right, we've already shown noisy latents before so we probably don't need to show them again.